tests/projects/test_projects_cli.py

```diff
@@ -29,3 +29,36 @@ def test_run_git():
             res = invoke_cli_runner(cli.run, [GIT_PROJECT_URI, "--no-conda", "-P", "alpha=0.5"])
             assert "python train.py 0.5 0.1" in res.output
             _assert_succeeded(res.output)
+
+GIT_SUBDIR_URI = "https://github.com/juntai-zheng/mlflow-git-features.git"
+
+
+def test_git_subdirectories():
+    with TempDir() as tmp:
+        with update_temp_env({mlflow.tracking._TRACKING_URI_ENV_VAR: tmp.path()}):
+            res = invoke_cli_runner(cli.run, [GIT_SUBDIR_URI + "#example/", "-e", "main", 
+                                              "--no-conda", "-P", "alpha=0.5"])
+            assert "python train.py 0.5 0.1" in res.output
+            _assert_succeeded(res.output)
+
+
+def test_missing_subdir():
+    with TempDir() as tmp:
+        with update_temp_env({mlflow.tracking._TRACKING_URI_ENV_VAR: tmp.path()}):
+            res = None
+            try:
+                res = invoke_cli_runner(cli.run, [GIT_PROJECT_URI + "#fake", 
+                                                  "--no-conda", "-P", "alpha=0.5"])
+            except AssertionError as e:
+                assert not res
+
+def test_restrict_periods():
+    with TempDir() as tmp:
+        with update_temp_env({mlflow.tracking._TRACKING_URI_ENV_VAR: tmp.path()}):
+            res = None
```

Comment:
You can use `with pytest.raises(...)` to simplify testing that an exception is raised